File: Assets/Scripts/Combat/BossGenerator.cs
using System;
using System.Collections.Generic;
using DokkaebiHand.Core;

namespace DokkaebiHand.Combat
{
    /// <summary>
    /// 보스 랜덤 생성기.
    /// 기본 도깨비 + 파츠 조합으로 매번 다른 보스를 생성.
    /// </summary>
    public class BossGenerator
    {
        private readonly Random _rng;

        public BossGenerator(int? seed = null)
        {
            _rng = seed.HasValue ? new Random(seed.Value) : new Random();
        }

        /// <summary>
        /// 스토리 보스 생성 (나선 1: 고정 도깨비 + 파츠 없음)
        /// </summary>
        public GeneratedBoss GenerateStoryBoss(int realmInSpiral, SpiralManager spiral)
        {
            var baseBoss = BossDatabase.GetBoss(Math.Min(realmInSpiral - 1, BossDatabase.AllBosses.Count - 1));
            int partsCount = spiral.GetPartsCount();
            var minRarity = spiral.GetMinPartsRarity();

            var parts = GenerateRandomParts(partsCount, minRarity);
            int target = spiral.GetTargetScore(baseBoss.TargetScore);

            // 파츠 목표 보너스 적용
            foreach (var part in parts)
            {
                target = (int)(target * (1f + part.TargetBonusPercent / 100f));
            }

            return new GeneratedBoss
            {
                BaseBoss = baseBoss,
                Parts = parts,
                FinalTargetScore = target,
                DisplayName = BuildDisplayName(baseBoss.NameKR, parts),
                Spiral = spiral.CurrentSpiral,
                AbsoluteRealm = spiral.AbsoluteRealm
            };
        }

        /// <summary>
        /// 잡졸 보스 생성 (완전 랜덤)
        /// </summary>
        public GeneratedBoss GenerateRandomBoss(SpiralManager spiral, int baseTarget = 150)
        {
            var allBosses = BossDatabase.AllBosses;
            var baseBoss = allBosses[_rng.Next(allBosses.Count)];
            int partsCount = spiral.GetPartsCount();
            var minRarity = spiral.GetMinPartsRarity();

            var parts = GenerateRandomParts(partsCount, minRarity);
            int target = spiral.GetTargetScore(baseTarget);

            foreach (var part in parts)
            {
                target = (int)(target * (1f + part.TargetBonusPercent / 100f));
            }

            // 기믹 간격 랜덤 변동 (±1, 최소 1)
            var mutatedBoss = new BossDefinition
            {
                Id = baseBoss.Id,
                Name = baseBoss.Name,
                NameKR = baseBoss.NameKR,
                Description = baseBoss.Description,
                TargetScore = baseBoss.TargetScore,
                Rounds = baseBoss.Rounds + _rng.Next(-1, 2), // 2~4 라운드 변동
                Gimmick = baseBoss.Gimmick,
                GimmickInterval = System.Math.Max(1, baseBoss.GimmickInterval + _rng.Next(-1, 2)),
                IntroDialogue = baseBoss.IntroDialogue,
                DefeatDialogue = baseBoss.DefeatDialogue,
                VictoryDialogue = baseBoss.VictoryDialogue,
                YeopReward = baseBoss.YeopReward + _rng.Next(-20, 30),
                DropsLegendaryTalisman = baseBoss.DropsLegendaryTalisman
            };
            mutatedBoss.Rounds = System.Math.Max(2, mutatedBoss.Rounds);

            return new GeneratedBoss
            {
                BaseBoss = mutatedBoss,
                Parts = parts,
                FinalTargetScore = target,
                DisplayName = BuildDisplayName(mutatedBoss.NameKR, parts),
                Spiral = spiral.CurrentSpiral,
                AbsoluteRealm = spiral.AbsoluteRealm
            };
        }

        private List<BossPartData> GenerateRandomParts(int count, PartsRarity minRarity)
        {
            var result = new List<BossPartData>();
            if (count <= 0) return result;

            var slots = new[] { PartsSlot.Head, PartsSlot.Arm, PartsSlot.Body };

            for (int i = 0; i < Math.Min(count, slots.Length); i++)
            {
                var pool = BossPartsDatabase.GetBySlotAndRarity(slots[i], minRarity);
                if (pool.Count > 0)
                {
                    result.Add(pool[_rng.Next(pool.Count)]);
                }
            }

            return result;
        }

        private string BuildDisplayName(string baseName, List<BossPartData> parts)
        {
            if (parts.Count == 0) return baseName;

            // 세트 효과 체크
            string setName = CheckSetBonus(parts);
            if (setName != null)
                return $"{setName} {baseName}";

            // 파츠 접두사 조합
            var prefixes = new List<string>();
            foreach (var part in parts)
            {
                prefixes.Add(part.NameKR);
            }

            return $"{string.Join(" ", prefixes)} {baseName}";
        }

        private string CheckSetBonus(List<BossPartData> parts)
        {
            var setCounts = new Dictionary<string, int>();
            foreach (var part in parts)
            {
                if (string.IsNullOrEmpty(part.SetId)) continue;
                if (!setCounts.ContainsKey(part.SetId))
                    setCounts[part.SetId] = 0;
                setCounts[part.SetId]++;
            }

            foreach (var kv in setCounts)
            {
                if (kv.Value >= 2)
                {
                    return kv.Key switch
                    {
                        "fire" => "불의 군주",
                        "ice" => "얼음의 군주",
                        "shadow" => "그림자의 군주",
                        "skull" => "해골의 군주",
                        _ => null
                    };
                }
            }

            return null;
        }
    }

    /// <summary>
    /// 생성된 보스 인스턴스
    /// </summary>
    public class GeneratedBoss
    {
        public BossDefinition BaseBoss;
        public List<BossPartData> Parts;
        public int FinalTargetScore;
        public string DisplayName;
        public int Spiral;
        public int AbsoluteRealm;

        public bool HasSetBonus()
        {
            var sets = new Dictionary<string, int>();
            foreach (var p in Parts)
            {
                if (string.IsNullOrEmpty(p.SetId)) continue;
                if (!sets.ContainsKey(p.SetId)) sets[p.SetId] = 0;
                sets[p.SetId]++;
                if (sets[p.SetId] >= 2) return true;
            }
            return false;
        }

        public float GetTotalTalismanReduction()
        {
            float total = 0;
            foreach (var p in Parts)
                total += p.TalismanEffectReduction;
            return Math.Min(total, 0.9f);
        }
    }
}

File: dokkaebi-love2d/src/combat/boss_generator.lua
--- 보스 랜덤 생성기
--- Ported from BossGenerator.cs (simplified — no BossParts system yet)

local BossData = require("src.combat.boss_data")

local PartsSlot = {
    Head = "head",
    Arm = "arm",
    Body = "body",
}

local PartsRarity = {
    Common = 1,
    Rare = 2,
    Legendary = 3,
}

--- 간소화된 파츠 데이터
local BossPartData = {}
BossPartData.__index = BossPartData

function BossPartData.new(t)
    return setmetatable({
        id = t.id or "",
        name = t.name or "",
        name_kr = t.name_kr or "",
        slot = t.slot or PartsSlot.Head,
        rarity = t.rarity or PartsRarity.Common,
        description = t.description or "",
        set_id = t.set_id or nil,
        target_bonus_percent = t.target_bonus_percent or 0,
        talisman_effect_reduction = t.talisman_effect_reduction or 0,
    }, BossPartData)
end

--- 생성된 보스 인스턴스
local GeneratedBoss = {}
GeneratedBoss.__index = GeneratedBoss

function GeneratedBoss.new(t)
    return setmetatable({
        base_boss = t.base_boss,
        parts = t.parts or {},
        final_target_score = t.final_target_score or 0,
        display_name = t.display_name or "",
        spiral = t.spiral or 1,
        absolute_realm = t.absolute_realm or 1,
    }, GeneratedBoss)
end

function GeneratedBoss:has_set_bonus()
    local sets = {}
    for _, p in ipairs(self.parts) do
        if p.set_id then
            sets[p.set_id] = (sets[p.set_id] or 0) + 1
            if sets[p.set_id] >= 2 then return true end
        end
    end
    return false
end

function GeneratedBoss:get_total_talisman_reduction()
    local total = 0
    for _, p in ipairs(self.parts) do
        total = total + p.talisman_effect_reduction
    end
    return math.min(total, 0.9)
end

-- ============================================================
-- BossGenerator
-- ============================================================
local BossGenerator = {}
BossGenerator.__index = BossGenerator

function BossGenerator.new(seed)
    if seed then math.randomseed(seed) end
    return setmetatable({}, BossGenerator)
end

--- 랜덤 보스 생성 (simplified)
--- spiral_manager: { current_spiral, absolute_realm, get_target_score(base), get_parts_count(), get_min_parts_rarity() }
function BossGenerator:generate_random_boss(spiral_manager, base_target)
    base_target = base_target or 150

    local all_bosses = BossData.get_all_bosses()
    local base_boss = all_bosses[math.random(#all_bosses)]

    -- 기본 목표 점수 (나선 배수 적용)
    local target = base_target
    if spiral_manager and spiral_manager.get_target_score then
        target = spiral_manager:get_target_score(base_target)
    end

    -- 기믹 간격 변동 (+-1, 최소 1)
    local interval = math.max(1, base_boss.gimmick_interval + math.random(-1, 1))
    local rounds = math.max(2, base_boss.rounds + math.random(-1, 1))
    local yeop = base_boss.yeop_reward + math.random(-20, 29)

    -- 변이된 보스 정의
    local mutated = BossData.BossDefinition({
        id = base_boss.id,
        name = base_boss.name,
        name_kr = base_boss.name_kr,
        description = base_boss.description,
        target_score = target,
        rounds = rounds,
        gimmick = base_boss.gimmick,
        gimmick_interval = interval,
        intro_dialogue = base_boss.intro_dialogue,
        defeat_dialogue = base_boss.defeat_dialogue,
        victory_dialogue = base_boss.victory_dialogue,
        yeop_reward = yeop,
        drops_legendary_talisman = base_boss.drops_legendary_talisman,
    })

    local spiral_num = 1
    local realm_num = 1
    if spiral_manager then
        spiral_num = spiral_manager.current_spiral or 1
        realm_num = spiral_manager.absolute_realm or 1
    end

    return GeneratedBoss.new({
        base_boss = mutated,
        parts = {},
        final_target_score = target,
        display_name = mutated.name_kr,
        spiral = spiral_num,
        absolute_realm = realm_num,
    })
end

--- 스토리 보스 생성 (고정 도깨비)
function BossGenerator:generate_story_boss(realm_in_spiral, spiral_manager)
    local index = math.min(realm_in_spiral, #BossData.get_all_bosses())
    local base_boss = BossData.get_boss(index)

    local target = base_boss.target_score
    if spiral_manager and spiral_manager.get_target_score then
        target = spiral_manager:get_target_score(base_boss.target_score)
    end

    local spiral_num = 1
    local realm_num = 1
    if spiral_manager then
        spiral_num = spiral_manager.current_spiral or 1
        realm_num = spiral_manager.absolute_realm or 1
    end

    return GeneratedBoss.new({
        base_boss = base_boss,
        parts = {},
        final_target_score = target,
        display_name = base_boss.name_kr,
        spiral = spiral_num,
        absolute_realm = realm_num,
    })
end

-- Export
BossGenerator.PartsSlot = PartsSlot
BossGenerator.PartsRarity = PartsRarity
BossGenerator.BossPartData = BossPartData
BossGenerator.GeneratedBoss = GeneratedBoss

return BossGenerator
